sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ReferenceToExpressions.scala

```diff
@@ -63,15 +63,33 @@ case class ReferenceToExpressions(result: Expression, children: Seq[Expression])
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val childrenGen = children.map(_.genCode(ctx))
-    val childrenVars = childrenGen.zip(children).map {
-      case (childGen, child) => LambdaVariable(childGen.value, childGen.isNull, child.dataType)
-    }
+    val (childrenVars, classChildrenVars) = childrenGen.zip(children).map {
+      case (childGen, child) =>
+        val childVar = LambdaVariable(childGen.value, childGen.isNull, child.dataType)
+
+        // SPARK-18125: The children vars are local variables. If the result expression uses
+        // splitExpression, those variables cannot be accessed so compilation fails.
+        // To fix it, we use class variables to hold those local variables.
+        val classChildVarName = ctx.freshName("classChildVar")
+        val classChildVarIsNull = ctx.freshName("classChildVarIsNull")
+        ctx.addMutableState(ctx.javaType(childVar.dataType), classChildVarName, "")
+        ctx.addMutableState("boolean", classChildVarIsNull, "")
+        val classChildVar =
+          LambdaVariable(classChildVarName, classChildVarIsNull, childVar.dataType)
+
+        (childVar, classChildVar)
+    }.unzip
+
+    val initClassChildrenVars = classChildrenVars.zipWithIndex.map { case (classChildrenVar, i) =>
```

Comment:
Just want to avoid tangling all codes together. I made the change to generate the code in previous block. Please see if it is better or not.